### PR TITLE
Adding Kinds to Type Automata

### PR DESCRIPTION
**Pull request by MarcoTz**

### CODE REVIEW

**Comment by MarcoTz** [comment]
This PR should only be merged after kind inference is done, as there are multiple workarounds needed to make sure the tests don't crash if the correct kinds are not annotated

File: src/TypeAutomata/FromAutomaton.hs [hunk 1]
```diff
@@ -110,28 +122,40 @@ argNodesToArgTypes argNodes rep = do
...
       (Cns, ns) -> do
          typs <- forM ns (nodeToType (flipPolarityRep rep))
+         let knd = checkTypKinds typs
          pure $ PrdCnsType CnsRep $ case rep of
-                                       PosRep -> mkInter defaultLoc defaultKind typs
-                                       NegRep -> mkUnion defaultLoc defaultKind typs
+                                       PosRep -> mkInter defaultLoc knd typs
+                                       NegRep -> mkUnion defaultLoc knd typs
+
+checkTypKinds :: [Typ pol] -> MonoKind
```
Comment (BinderDavid) [inline src/TypeAutomata/FromAutomaton.hs:136]: Throw monadic errors in this function, not error thunks.

File: src/TypeAutomata/ToAutomaton.hs [hunk 2]
```diff
@@ -26,6 +26,9 @@ import Loc ( defaultLoc )
 import Utils ( enumerate )
 import Control.Monad
 
+--- default Kind where type is not yet clear
```
Comment (BinderDavid) [inline src/TypeAutomata/ToAutomaton.hs:29]: The occurrence of `defaultKind` is definitely sus ;)

I think you should have enough information in the input to never use the defaultKind here?

File: src/TypeAutomata/Determinize.hs [hunk 3]
```diff
@@ -80,19 +80,24 @@ combineNodeLabels (fstLabel@MkNodeLabel{}:rs) =
   case rs_merged of
     pr@MkPrimitiveNodeLabel{} -> error ("Tried to combine primitive type" <> show pr <> " and algebraic type " <> show fstLabel)
     combLabel@MkNodeLabel{} ->
```
Comment (BinderDavid) [inline src/TypeAutomata/Determinize.hs:82]: Es gibt in dieser Funktion keinen Test der auf Gleichheit von Kinds testet? (Zusätzlich zu dem Test der Gleichheit von Polarity).

File: src/TypeAutomata/FromAutomaton.hs (hunk 1 above)
Comment (MarcoTz) [reply src/TypeAutomata/FromAutomaton.hs]: Changed it to monadic errors

**Review by BinderDavid — APPROVED** [review]

**Pull request merged**